### PR TITLE
[libc][Docs] Update the GPU RPC documentation

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
This adds some more concrete information on the RPC interface. Hopefully
this is intelligable and provides some useful examples.

### CODE REVIEW

**Review by lntue — APPROVED** [review]

File: libc/docs/gpu/rpc.rst [hunk 1]
```diff
@@ -11,10 +11,298 @@ Remote Procedure Calls
...
+with the platform's operating system through system calls. The GPU however does 
+not provide an operating system that can handle target dependent operations.
+Instead, we implemented remote procedure calls to interface with the host's 
+operating system while executing on a GPU.
+
+We implemented remote procedure calls using unified virtual memory to create a 
+shared communicate channel between the two processes. This memory is often 
+pinned memory that can be accessed asynchronously and atomically by multiple 
+processes simultaneously. This supports means that we can simply provide mutual 
+exclusion on a shared better to swap work back and forth between the host system
```
Comment (nickdesaulniers) [inline libc/docs/gpu/rpc.rst:24]: s/better/buffer/

File: libc/docs/gpu/rpc.rst [hunk 2]
```diff
@@ -11,10 +11,298 @@ Remote Procedure Calls
...
+We implemented remote procedure calls using unified virtual memory to create a 
+shared communicate channel between the two processes. This memory is often 
+pinned memory that can be accessed asynchronously and atomically by multiple 
+processes simultaneously. This supports means that we can simply provide mutual 
+exclusion on a shared better to swap work back and forth between the host system 
+and the GPU. We can then use this to create a simple client-server protocol 
+using this shared memory.
+
+This work treats the GPU as a client and the host as a server. The client 
+initiates a communication while the server listens for them. In order to
```
Comment (nickdesaulniers) [inline libc/docs/gpu/rpc.rst:29]: s/them/requests/

File: libc/docs/gpu/rpc.rst [hunk 3]
```diff
@@ -11,10 +11,298 @@ Remote Procedure Calls
...
+
+* The server searches for an available port with pending work and claims the 
+  lock.
+* The server checks that the port is still available to the current device.
+* The server reads the opcode to perform the expected operation, in this 
+  case a receive and then send.
+* The server reads the data from the fixed-size packet.
+* The server writes its data to the fixed-size packet and toggles its outbox.
+* The server closes the port and continues searching for ports that need to be 
+  serviced
```
Comment (nickdesaulniers) [inline libc/docs/gpu/rpc.rst:86]: add punctuation

File: libc/docs/gpu/rpc.rst [hunk 4]
```diff
@@ -11,10 +11,298 @@ Remote Procedure Calls
...
+
+The above code must be compiled in CUDA's relocatable device code mode and with 
+the advanced offloading driver to link in the library. Currently this can be 
+done with the following invocation. Using LTO avoids the overhead normally 
+associated with relocatable device code linking.
+
+.. code-block:: sh
+
+  $> clang++ -x cuda rpc.cpp --offload-arch=native -fgpu-rdc -lcudart -lcgpu \
+       -I<install-path>include -L<install-path>/lib -lllvmlibc_rpc_server \
```
Comment (nickdesaulniers) [inline libc/docs/gpu/rpc.rst:302]: add `/` before `include` for the `-I` to be consistent with `-L`

**Review by nickdesaulniers — COMMENTED** [review]
Thanks for the write up! This was helpful

**Pull request merged**